src/crawlee/service_locator.py

```diff
@@ -0,0 +1,147 @@
+from __future__ import annotations
+
+from crawlee._utils.docs import docs_group
+from crawlee.base_storage_client._base_storage_client import BaseStorageClient
+from crawlee.configuration import Configuration
+from crawlee.errors import ServiceConflictError
+from crawlee.events._event_manager import EventManager
+
+__all__ = list[str]()  # Prefer "from crawlee import service_locator"
+
+
+class _ServiceLocator:
+    """Service locator for managing the services used by Crawlee.
+
+    All services are initialized to its default value lazily.
+    """
+
+    def __init__(self) -> None:
+        self._configuration: Configuration | None = None
+        self._event_manager: EventManager | None = None
+        self._storage_client: BaseStorageClient | None = None
+
+        # Flags to check if the services were already set.
+        self._configuration_was_set = False
+        self._event_manager_was_set = False
+        self._storage_client_was_set = False
+
+    @property
+    def configuration(self) -> Configuration:
+        if self._configuration is None:
+            self._configuration = Configuration()
+
+        return self._configuration
+
+    @configuration.setter
+    def configuration(self, value: Configuration) -> None:
+        self._configuration = value
+        self._configuration_was_set = True
+
+    @property
+    def storage_client(self) -> BaseStorageClient:
+        if self._storage_client is None:
+            from crawlee.memory_storage_client import MemoryStorageClient
+
+            self._storage_client = MemoryStorageClient()
+
+        return self._storage_client
+
+    @storage_client.setter
+    def storage_client(self, value: BaseStorageClient) -> None:
+        self._storage_client = value
+        self._storage_client_was_set = True
+
+    @property
+    def event_manager(self) -> EventManager:
+        if self._event_manager is None:
+            from crawlee.events import LocalEventManager
+
+            self._event_manager = LocalEventManager()
+
+        return self._event_manager
+
+    @event_manager.setter
+    def event_manager(self, value: EventManager) -> None:
+        self._event_manager = value
+        self._event_manager_was_set = True
+
+    @property
+    def configuration_was_set(self) -> bool:
+        return self._configuration_was_set
+
+    @property
+    def event_manager_was_set(self) -> bool:
+        return self._event_manager_was_set
+
+    @property
+    def storage_client_was_set(self) -> bool:
+        return self._storage_client_was_set
+
+
+_service_locator = _ServiceLocator()
```

Comment:
We discussed that we could just re-export this as `crawlee.service_locator` and rename this module to `crawlee._service_locator`. That way, we won't need the functions below.

Reply:
Done, much better, thanks.